ztoc/ztoc.go

```diff
@@ -85,13 +89,24 @@ type FileMetadata struct {
 	Xattrs map[string]string
 }
 
-// FileExtractConfig contains information used to extract a file from compressed data.
-type FileExtractConfig struct {
-	UncompressedSize      compression.Offset
-	UncompressedOffset    compression.Offset
-	Checkpoints           []byte
-	CompressedArchiveSize compression.Offset
-	MaxSpanID             compression.SpanID
+// FileMode gets file mode for the file metadata
+func (src FileMetadata) FileMode() (m os.FileMode) {
+	// FileMetadata.Mode is tar.Header.Mode so we can understand the these bits using `tar` pkg.
+	m = (&tar.Header{Mode: src.Mode}).FileInfo().Mode() &
+		(os.ModePerm | os.ModeSetuid | os.ModeSetgid | os.ModeSticky)
+	switch src.Type {
+	case "dir":
+		m |= os.ModeDir
+	case "symlink":
+		m |= os.ModeSymlink
+	case "char":
+		m |= os.ModeDevice | os.ModeCharDevice
+	case "block":
+		m |= os.ModeDevice
+	case "fifo":
+		m |= os.ModeNamedPipe
```

Comment:
Can this exist as a map defined outside the function, and then a lookup in the function, instead of this switch?